File: Problem_2 Valid_Sudoku.py
# // Time Complexity : O(9*9)
# // Space Complexity : O(9)
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : No
#
#
# // Your code here along with comments explaining your approach
def notInRow(arr, row):
    st = set()
    for i in range(0, 9):
        if arr[row][i] in st:
            return False

        if arr[row][i] != '.':
            st.add(arr[row][i])
    return True

def notInCol(arr, col):
    st = set()

    for i in range(0, 9):
        if arr[i][col] in st:
            return False
        if arr[i][col] != '.':
            st.add(arr[i][col])
    return True

def notInBox(arr, startRow, startCol):
    st = set()

    for row in range(0, 3):
        for col in range(0, 3):
            curr = arr[row + startRow][col + startCol]
            if curr in st:
                return False
            if curr != '.':
                st.add(curr)

    return True

def isValid(arr, row, col):
    return (notInRow(arr, row) and notInCol(arr, col) and
            notInBox(arr, row - row % 3, col - col % 3))


def isValidConfig(arr, n):
    for i in range(0, n):
        for j in range(0, n):
            if not isValid(arr, i, j):
                return False

    return True



if __name__ == "__main__":

    board = [['5', '3', '.', '.', '7', '.', '.', '.', '.'],
             ['6', '.', '.', '1', '9', '5', '.', '.', '.'],
             ['.', '9', '8', '.', '.', '.', '.', '6', '.'],
             ['8', '.', '.', '.', '6', '.', '.', '.', '3'],
             ['4', '.', '.', '8', '.', '3', '.', '.', '1'],
             ['7', '.', '.', '.', '2', '.', '.', '.', '6'],
             ['.', '6', '.', '.', '.', '.', '2', '8', '.'],
             ['.', '.', '.', '4', '1', '9', '.', '.', '5'],
             ['.', '.', '.', '.', '8', '.', '.', '7', '9']]

    if isValidConfig(board, 9):
        print("True")
    else:
        print("False")

File: Problem2_ValidSudoku.java
// Time Complexity : o(n2)
// Space Complexity : o(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
class Solution {
    public boolean isValidSudoku(char[][] board) {
        
        if(board==null|| board.length==0)
            return true;
        
        int m=board.length;
        int n=board[0].length;
        
        // for row
        for(int i=0;i<m;i++)
        {
            HashSet<Character> c=new HashSet<>();
            for(int j=0;j<n;j++)
            {
                if(board[j][i]!='.')
                {
                    if(!c.contains(board[j][i]))
                        c.add(board[j][i]);
                    else
                        return false;
                }
            }
        }
        
        // for column
        for(int i=0;i<m;i++)
        {
            HashSet<Character> c=new HashSet<>();
            for(int j=0;j<n;j++)
            {
                if(board[i][j]!='.')
                {
                    if(!c.contains(board[i][j]))
                        c.add(board[i][j]);
                    else
                        return false;
                }
            }
        }
        
        // for 3*3 
         
         for(int i = 0; i < 9; i= i + 3) {
            for(int j = 0; j < 9; j = j + 3) {
                HashSet<Character> c = new HashSet<>();
                for(int x = i; x < i + 3; x++) {
                    for(int y = j; y < j + 3; y++) {
	  //mat is the name of the sudoku matrix
                char currchar = board[x][y];
	    //when current cell in sudoku is not "." empty
                if(currchar != '.') {
            //check if hashset already contains currchar, return false if it
	 //is already intialized in set
	     if (c.contains(currchar))
                    return false;
                    //else add it to set
                    c.add(currchar);
                    }
                }
            }
        }
    }
        return true;
    }
}
